inout_test.go

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fx_test
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/dig"
+	"go.uber.org/fx"
+	"go.uber.org/fx/fxtest"
+)
+
+func TestIn(t *testing.T) {
+	t.Parallel()
+
+	type in struct {
+		fx.In
+	}
+	assert.True(t, dig.IsIn(reflect.TypeOf(in{})), "Expected dig.In to work with fx.In")
+}
+
+func TestOut(t *testing.T) {
+	t.Parallel()
+
+	type out struct {
+		fx.Out
+	}
+	assert.True(t, dig.IsOut(reflect.TypeOf(out{})), "expected dig.Out to work with fx.Out")
+}
+
+func TestOptionalTypes(t *testing.T) {
+	t.Parallel()
+
+	type foo struct{}
+	newFoo := func() *foo { return &foo{} }
+
+	type bar struct{}
+	newBar := func() *bar { return &bar{} }
+
+	type in struct {
+		fx.In
+
+		Foo *foo
+		Bar *bar `optional:"true"`
+	}
+
+	t.Run("NotProvided", func(t *testing.T) {
+		t.Parallel()
+
+		ran := false
+		app := fxtest.New(t, fx.Provide(newFoo), fx.Invoke(func(in in) {
+			assert.NotNil(t, in.Foo, "foo was not optional and provided, expected not nil")
+			assert.Nil(t, in.Bar, "bar was optional and not provided, expected nil")
+			ran = true
+		}))
+		app.MustStart().MustStop()
+		assert.True(t, ran, "expected invoke to run")
+	})
+
+	t.Run("Provided", func(t *testing.T) {
+		t.Parallel()
+
+		ran := false
+		app := fxtest.New(t, fx.Provide(newFoo, newBar), fx.Invoke(func(in in) {
+			assert.NotNil(t, in.Foo, "foo was not optional and provided, expected not nil")
+			assert.NotNil(t, in.Bar, "bar was optional and provided, expected not nil")
+			ran = true
+		}))
+		app.MustStart().MustStop()
+		assert.True(t, ran, "expected invoke to run")
+	})
+}
+
+func TestNamedTypes(t *testing.T) {
+	t.Parallel()
+
+	type a struct {
+		name string
+	}
```

Comment:
I found this test quite confusing. We're testing named fields, but the fact that the field is named `name` is irrelevant. However, the `name:"..."` tag *is* relevant. Can we give the field a different name, so that it's clearer that only the tag matters?